### PR TITLE
KAFKA-4850:RocksDb cannot use Bloom Filters

### PR DESCRIPTION
**Pull request by bharatviswa504**
Added BloomFilter to speedup rocksdb lookup.

### CODE REVIEW

**Comment by bharatviswa504** [comment]
@enothereska @mjsax  @guozhangwang Could you please review.

**Comment by bharatviswa504** [comment]
retest this please

**Comment by bharatviswa504** [comment]
tests passed locally.

**Comment by bharatviswa504** [comment]
The test failure on JDK 7 is happening on other PR's too.
Test failure is not related to this change.

**Comment by bharatviswa504** [comment]
retest this please

**Comment by enothereska** [comment]
Thank you. I think it looks good. I've started some system tests to see what impact this has on performance, will post link here once done. @bharatviswa504 do you know what the memory overhead might be when turning on bloom filters?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 1]
```diff
@@ -112,11 +113,14 @@
     @SuppressWarnings("unchecked")
     public void openDB(ProcessorContext context) {
         // initialize the default rocksdb options
+
+        options = new Options();
+
         final BlockBasedTableConfig tableConfig = new BlockBasedTableConfig();
         tableConfig.setBlockCacheSize(BLOCK_CACHE_SIZE);
         tableConfig.setBlockSize(BLOCK_SIZE);
+        tableConfig.setFilter(new BloomFilter(10));
```
Comment (dguy) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: What does the 10 mean? Could you extract it to a constant so it is named?

**Comment by enothereska** [comment]
system test passed: https://jenkins.confluent.io/job/system-test-kafka-branch-builder-2/291/console. I don't necessarily see a perf improvement, but I'm not sure the tests are designed for showing off any improvement. @bharatviswa504 any suggestions on a good test/benchmark?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java (hunk 1 above)
Comment (bharatviswa504) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: @dguy It is bits per key. Added a constant to name it.

**Comment by bharatviswa504** [comment]
@enothereska 
https://github.com/facebook/rocksdb/wiki/RocksDB-Bloom-Filter#memory-usage

Do you think increasing block size will improve performance?

I don't have much details with rocksdb, let me know if I am missing something.

**Review by dguy — APPROVED** [review]
LGTM - though would be good to understand how this impacts memory usage, performance etc

**Comment by guozhangwang** [comment]
My understanding is that bloom filters is only beneficial for single key-value lookups, so the current system tests may not be best fit for analyzing its impacts. More specifically, for windowed rocksdb store since we always store the window start timestamp as prefix of the key, which is 64bits, hence first 10bits bloom filter would not help much on filtering those keys; for such cases it is better to have the prefix seeking support as discussed here:

https://confluentinc.atlassian.net/wiki/display/Engineering/Discussion%3A+Non-key+KTable-KTable+Joins#Discussion:Non-keyKTable-KTableJoins-Option1:useRocksDBprefixseeking

The code itself LGTM, but I think it is better to execute a benchmark with the following settings before merging the PR:

1. have a running aggregate in Streams with RocksDBStore (i.e. not windowed store).
2. have another thread continuously query the rocksDB store on single key lookup. 
3. compare the query QPS with / wo this PR; expect to have non-neglectible perf diffs.

@bharatviswa504 Do you want to do this benchmark?

**Comment by mjsax** [comment]
Replaced by #6012

**Pull request closed without merging**